src/main/kotlin/org/rust/lang/core/types/infer/MemoryCategorization.kt

```diff
@@ -5,57 +5,167 @@
 
 package org.rust.lang.core.types.infer
 
+import com.intellij.psi.util.parentOfType
 import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsElement
 import org.rust.lang.core.psi.ext.containerExpr
 import org.rust.lang.core.psi.ext.mutability
 import org.rust.lang.core.types.builtinDeref
-import org.rust.lang.core.types.infer.Adjustment.Deref
+import org.rust.lang.core.types.infer.BorrowKind.ImmutableBorrow
+import org.rust.lang.core.types.infer.BorrowKind.MutableBorrow
+import org.rust.lang.core.types.infer.Categorization.*
+import org.rust.lang.core.types.infer.ImmutabilityBlame.*
+import org.rust.lang.core.types.infer.InteriorKind.InteriorElement
+import org.rust.lang.core.types.infer.InteriorKind.InteriorField
+import org.rust.lang.core.types.infer.MutabilityCategory.Declared
+import org.rust.lang.core.types.infer.PointerKind.BorrowedPointer
+import org.rust.lang.core.types.infer.PointerKind.UnsafePointer
 import org.rust.lang.core.types.inference
 import org.rust.lang.core.types.isDereference
+import org.rust.lang.core.types.regions.ReStatic
+import org.rust.lang.core.types.regions.Region
 import org.rust.lang.core.types.ty.*
 import org.rust.lang.core.types.type
 import org.rust.stdext.nextOrNull
 
 
+/** [Categorization] is a subset of the full expression forms */
+sealed class Categorization {
+    /** Temporary value */
+    data class Rvalue(val region: Region) : Categorization()
+
+    /** Static value */
+    object StaticItem : Categorization()
+
+    /** Variable captured by closure */
+    object Upvar : Categorization()
+
+    /** Local variable */
+    data class Local(val element: RsElement) : Categorization()
+
+    /** Dereference of a pointer */
+    data class Deref(val cmt: Cmt, val pointerKind: PointerKind?) : Categorization()
+
+    /** Something reachable from the base without a pointer dereference (e.g. field) */
+    data class Interior(val cmt: Cmt, val interiorKind: InteriorKind) : Categorization()
+
+    /** Selects a particular enum variant (if enum has more than one variant */
+    data class Downcast(val cmt: Cmt, val element: RsElement) : Categorization()
+}
+
+sealed class BorrowKind {
```

Comment:
Maybe `enum` is better suited in this case?

Reply:
Maybe, but I use `sealed class` to ensure uniformity (see `PointerKind`)